### PR TITLE
[hdlc] update buffer model for Encoder and Decoder

### PR DESCRIPTION
**Pull request by abtink**
This PR contains two related commits: 

**[hdlc] update buffer model for Encoder and Decoder**
    
This commit updates HDLC implementation to harmonize the buffer model used by `Hdlc::Encoder` and `Hdlc::Decoder`.

The new model defines a common class `Hdlc::FrameWritePointer` which provides a minimum set of APIs used by `Encoder/Decoder` for writing frames, while not defining the underlying buffer space or how the buffer is managed. `Encoder` or `Decoder` users are expected to use sub-classes of `FrameWritePointer` to add the buffer space and implement the frame buffer management scheme.

Two sub-classes `Hdlc::FrameBuffer` and `Hdlc::MultiFrameBuffer` are also provided which respectively allow storing a single frame or
multiple frames (FIFO queue of frames) in a buffer of given size. 

This commit also updates the `NcpUart` code to use the new buffer model.

Finally, this commit adds an HDLC unit test `test_hdlc.cpp` which covers the following:

- Verify behavior of `Hdlc::FrameBuffer` (single frame buffer),
- Verify behavior of `Hdlc::MultiFrameBuffer` (FIFO queue),
- Verify behavior of `Hdlc::Encoder` and `Hdlc::Decoder`,
- Test encoder/decoder with randomly generated frames (fuzz test).

----

**[posix-app] adopt new HDLC buffer models**
    
This commit updates the `HdlcInterface` and `RadioSpinel` to use the new HDLC buffer model. In particular, for received spinel frames the decoder uses an `Hdlc::MultiFrameBuffer` which is capable of storing multiple frames in a FIFO queue manner.

With this model, the received and decoded spinel frames are directly placed in the FIFO buffer. This allows `RadioSpinel` code during `WaitResponse()` (when waiting for specific Spinel response frame) to simply save/keep a notification frame in the queue buffer (without requiring to copy the frame) so that it can read and process the saved frame later. If a received frame can be processed at the time, the frame is then simply discarded from the queue  buffer.

### CODE REVIEW

**Comment by bukepo** [comment]
Thanks. Also remove frame_queue?

File: src/ncp/hdlc.hpp [hunk 1]
```diff
@@ -51,107 +54,390 @@ namespace ot {
 namespace Hdlc {
 
 /**
- * This class implements the HDLC-lite encoder.
+ * This class defines a frame write pointer used by `Hdlc:Encoder` or `Hdlc:Decoder`.
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: nit: "Hdlc:Encoder" -> "Hdlc::Encoder" and same for Decoder?

File: src/ncp/hdlc.hpp [hunk 2]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+ *
+ * This class defines the minimum set of APIs used by `Encoder/Decoder` for writing an encoded/decoded frame. It is
+ * simply a wrapper over a pointer into a buffer indicating where next byte should be written. Along with a write
+ * pointer, this class stores a remaining length variable indicating number of remaining bytes that can be written into
+ * the buffer.
+ *
+ * Note that this class does NOT define the underlying buffer space or how it is being managed.
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: nit: use `@note`?

File: src/ncp/hdlc.hpp [hunk 3]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+    otError WriteByte(uint8_t aByte)
+    {
+        return CanWrite(sizeof(uint8_t)) ? (*mWritePointer++ = aByte, mRemainingLength--, OT_ERROR_NONE)
+                                         : OT_ERROR_NO_BUFS;
+    }
+
+    /**
+     * This method undoes the last @p aUndoLength writes, removing them from frame.
+     *
+     * Note that caller should ensure that @p aUndoLength is less than or equal number of previously written bytes into
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: nit: use `@note`?

File: src/ncp/hdlc.hpp [hunk 4]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+ * This class defines a template frame buffer of a given size for storing a single frame.
+ *
+ * The template parameter `kSize` specifies the size of the buffer.
+ *
+ */
+template <uint16_t kSize> class FrameBuffer : public FrameWritePointer
+{
+public:
+    /**
+     * This constructor initializes the `FrameBuffer` object`
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: nit: extra back-tick at end of line

File: src/ncp/hdlc.hpp [hunk 5]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+    /**
+     * This method gets a pointer to start of the current frame.
+     *
+     * @returns A pointer to start of the frame.
+     *
+     */
+    uint8_t *GetFrame(void) { return mWriteFrameStart + kHeaderSize; }
+
+    /**
+     * This method saves the current frame. It also prepares the write pointer for a next frame to be written into the
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: nit: first line brief should be single sentence

File: src/ncp/hdlc.hpp [hunk 6]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+    void SaveFrame(void)
+    {
+        Encoding::LittleEndian::WriteUint16(GetLength(), mWriteFrameStart);
+        mWriteFrameStart = mWritePointer;
+        mWritePointer += kHeaderSize;
+        mRemainingLength = static_cast<uint16_t>(mBuffer + kSize - mWritePointer);
+    }
+
+    /**
+     * This method discards the current frame. It also resets the write pointer for a next frame.
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: nit: first line brief should be single sentence

File: src/ncp/hdlc.hpp [hunk 7]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+
+private:
+    /*
+     * The diagram below illustrates how the frames are saved in the buffer.
+     *
+     * Each saved frame contains a header which is 2 bytes long and specifies the frame length (the length does not
+     * include the header itself). The frame length is stored in header bytes as a `uint16_t` value using little-endian
+     * encoding.
+     *
+     * The digram shows `mBuffer` and different pointers to the buffer. It represent buffer state when there are two
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: typo: "digram" -> "diagram"

File: src/ncp/hdlc.hpp [hunk 8]
```diff
@@ -51,107 +54,390 @@ namespace ot {
...
+{
+public:
     /**
      * This constructor initializes the object.
      *
+     * @param[in] aWritePointer   The `FrameWritePointer` used by `Encoder` to write the encoded frames.
+     *
      */
-    Encoder(void);
+    Encoder(FrameWritePointer &aWritePointer);
```
Comment (jwhui) [inline src/ncp/hdlc.hpp]: Declare as `explicit`?

Comment (abtink) [reply src/ncp/hdlc.hpp]: All changed in the new pushed commit. Thanks.

**Pull request merged**